### ISSUE TITLE
scatter_reduce

### ISSUE THREAD

**Issue by cpuhrsch** [pytorch/pytorch#63780]
## 🚀 Feature

This feature requests defines scatter_reduce, its signatures and the steps to implement it as discussed in https://github.com/pytorch/pytorch/issues/22378. scatter_reduce is meant to satisfy the feature requests in https://github.com/pytorch/pytorch/issues/22378  under light of a recent [resurgence of demand](https://github.com/pytorch/pytorch/issues/22378#issuecomment-862705586).

## Motivation

Amongst others Graph Neural Networks need a scatter operation followed by a reduction. This demand has already manifested itself in the excellent library [torch_scatter](https://github.com/rusty1s/pytorch_scatter) and is used by the very popular library [pytorch_geometric](https://github.com/rusty1s/pytorch_geometric). At a minimum official support shifts the maintenance and release burden from one of our very busy community members @rusty1s onto PyTorch.

## Pitch

scatter_reduce matches the feature set and performance of [torch_scatters's scatter](https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html) with a slightly different signature.

### Signature

```scatter_reduce(input, dim, index, reduce, dim_size=None, out=None)```

```input``` - The source Tensor
```dim``` - The axis along which to index
```index``` - The indices of elements to scatter.
```reduce``` - Must be one of ```"sum"```, ```"prod"```, ```"mean"```, ```"amin"``` or ```"amax"```
```output_size``` - If specified defines the size of the output along dimension ```dim```.
```out``` - Optional output Tensor.

```index``` may broadcast with ```input``` and must match in size.

The order of these arguments differ from the signature of torch_scatters's scatter and matches that of [torch.Tensor.scatter_](https://pytorch.org/docs/master/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_) instead. This prevents a default value for dim, which torch_scatters's scatter sets to -1, but consistency here wins over saving on these few keystrokes. Further ```output_size``` is used instead of ```dim_size``` to be consistent with similar arguments in operators such as [repeat_interleave](https://pytorch.org/docs/master/generated/torch.repeat_interleave.html), [adaptive_avg_pool1d](https://pytorch.org/docs/master/generated/torch.nn.functional.adaptive_avg_pool1d.html) or [Fold](https://pytorch.org/docs/master/generated/torch.nn.Fold.html).

### Formal behavior
Given ```input``` and ```index``` have size ```(x_0, ..., x_{i -1}, x_{i}, x_{i + 1}, ... x{n - 1})``` and ```dim = i```.

If ```out``` is not specified it will be set to a Tensor of size ```(x_0, ..., x_{i - 1}, y, x_{i + 1}, ..., x_{n - 1}``` where ```y``` is ```(max_{j} index.reshape(-1)_j) + 1```. It is further initialized with the reduction's identity value, e.g. for min this would be the dtype's maximum value. If ```out``` is specified ```out.size(i)``` must be greater than or equal to ```y```.

```out``` is included in the reduction.

Specifically if ```src``` and ```input``` are 1-dimensional vectors and reduce is ```"sum"``` we have
```out_{index_j} == out_{index_j} + \sum_{j} src_{j}```.

if ```src``` and ```input``` are 1-dimensional vectors and reduce is ```"amin"``` we have
```out_{index_j} == min(out_{index_j},  \min_{j} src_{j})``` where ```index_j = i```.

```index``` may include repeat values. This means the corresponding element in ```out``` is updated atomically. There are no other constraints on ```index```.

```output_size``` can be used to specify a```y``` ahead of time. This can be used as a performance optimization, because otherwise a scan of ```index``` is required to infer the right size of the output.

There are no determinism guarantees for CUDA, albeit we may eventually introduce this behavior to ease debugging eventually.

### Implementation proposal

1. Implement minimal behavior. In particular ```input``` must match ```index``` in dimension, have dimension two and  ```reduce``` is constraint to ```"sum"``` and we only support CPU without specialized (e.g. vectorized) kernels. Includes lots of tests.
2. Add Autograd support.
3. Add documentation
4. Extend to basic behavior. In particular ```input``` must match ```index``` in dimension, but may be n dimension. ```reduce``` is extended to ```"sum"```, ```"prod"``` and ```"amin"``` (prior and subsequent operators enables the others) and we only support CPU without specialized (e.g. vectorized) kernels.
5. Support broadcasting.
6. Support CUDA.
7. Add benchmarks
8. Add specialized CPU code.
9. Support additional reductions.
10. (optional) Introduce a deterministic mode

## Alternatives

Don't do anything.

## Additional context

Graph Neural Networks also benefit from extended support for sparse operators.

**Comment by cpuhrsch** [comment]
Only the first step landed....

**Comment by Linux-cpp-lisp** [comment]
It looks like this exists now for `sum` and `multiply`; is support for `min`/`max`/`mean` still planned?

**Comment by cpuhrsch** [comment]
@Linux-cpp-lisp - yes it is.

**Comment by louisabraham** [comment]
Hi, what about logsumexp reduction? https://github.com/pytorch/pytorch/issues/31394

**Comment by tvercaut** [comment]
For those like me seeing this issue open (and being slightly [confused](https://github.com/pytorch/pytorch/issues/22378#issuecomment-1411211744)), it might be good to note that [`scatter_reduce`](https://pytorch.org/docs/1.13/generated/torch.Tensor.scatter_reduce_.html#torch.Tensor.scatter_reduce_) is already available and offers more reduction options ("sum", "prod", "mean", "amax", "amin") than [`scatter`](https://pytorch.org/docs/1.13/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_) ("add", "multiply"). I guess this issue remain open to account for other reduction modes not yet being supported.

**Comment by mikaylagawarecki** [comment]
@tvercaut Thanks for pointing this out, many apologies for the confusion. I'm closing as complete since scatter_reduce has been landed, will create a separate tracking issue for feature requests for other reduction modes.
